core/services/ocr2/plugins/ocr2keeper/evm21/registry_check_pipeline.go

```diff
@@ -361,6 +369,12 @@ func (r *EvmRegistry) simulatePerformUpkeeps(ctx context.Context, checkResults [
 			r.lggr.Warnf("upkeepId %s is not eligible after simulation of perform", checkResults[idx].UpkeepID.String())
 			checkResults[performToKeyIdx[i]].Eligible = false
 			checkResults[performToKeyIdx[i]].IneligibilityReason = uint8(encoding.UpkeepFailureReasonSimulationFailed)
+		} else {
+			// at this point, the core node knows the exact perform data of the upkeep and the call data to L1.
+			// it can calculate a relatively accurate L1 gas cost
```

Comment:
this may introduce some delays between when L1 gas cost is calculated and when users are charged.